{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analyze curated mutational counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import Python modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats\n",
    "import re\n",
    "from collections import defaultdict\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "import math\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "sns.set_theme(font_scale=1.0, style='ticks', palette='colorblind')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in counts data, compute fitness effects using old and new approaches, and read in data on gene boundaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in counts data and add columns giving truncated counts\n",
    "counts_df = pd.read_csv('../results/curated_mut_counts.csv')\n",
    "counts_df['mut_type_arrow'] = counts_df['mut_type'].apply(\n",
    "    lambda x: x[0]+r'$\\rightarrow$'+x[1]\n",
    ")\n",
    "counts_df['site'] = counts_df['nt_site']\n",
    "counts_df['ss_paired'] = counts_df['ss_prediction'] == 'paired'\n",
    "\n",
    "# Add new columns...\n",
    "actual_count_cols = [\n",
    "    'actual_count',\n",
    "    'actual_count_England',\n",
    "    'actual_count_USA',\n",
    "    'actual_count_pre_omicron',\n",
    "    'actual_count_omicron'\n",
    "]\n",
    "clipping_upper_lim = 0.98\n",
    "dfs = []\n",
    "for (mut_type, mut_type_data) in counts_df.groupby('mut_type'):\n",
    "    \n",
    "    # Add a column with actual counts truncated at the indicated percentile of the distribution\n",
    "    # of counts for synonymous mutations\n",
    "    syn_data = mut_type_data[mut_type_data['mut_class'] == 'synonymous'].copy()\n",
    "    for actual_count_col in actual_count_cols:\n",
    "        upper_lim = syn_data[actual_count_col].quantile(clipping_upper_lim)\n",
    "        syn_data[f'truncated_{actual_count_col}'] = syn_data[actual_count_col].clip(upper=upper_lim)\n",
    "        mut_type_data[f'truncated_{actual_count_col}'] = mut_type_data[actual_count_col].clip(upper=upper_lim)\n",
    "\n",
    "    # Add columns with expected counts computed from the curated counts data\n",
    "    mut_type_data['expected_count_mean'] = syn_data['actual_count'].mean()\n",
    "    mut_type_data['expected_count_mean_truncated'] = syn_data['truncated_actual_count'].mean()\n",
    "    mut_type_data['expected_count_median'] = syn_data['actual_count'].median()\n",
    "\n",
    "    # Add columns of counts shuffled between mutations within the same class (syn, nonsyn, nonsense, noncoding)\n",
    "    for (mut_class, mut_class_data) in mut_type_data.groupby('mut_class'):\n",
    "        mut_class_data['shuffled_actual_count'] = list(mut_class_data['actual_count'].sample(frac=1, random_state=1))\n",
    "        mut_class_data['shuffled_truncated_actual_count'] = list(mut_class_data['truncated_actual_count'].sample(frac=1, random_state=1))\n",
    "        dfs.append(mut_class_data)\n",
    "\n",
    "counts_df = pd.concat(dfs)\n",
    "\n",
    "# Compute fitness values using multiple different strategies\n",
    "counts_df['fitness'] = counts_df.apply(\n",
    "    lambda row: np.log((row['actual_count'] + 0.5) / (row['expected_count'] + 0.5)),\n",
    "    axis=1\n",
    ")\n",
    "counts_df['fitness_mean'] = counts_df.apply(\n",
    "    lambda row: np.log((row['actual_count'] + 0.5) / (row['expected_count_mean'] + 0.5)),\n",
    "    axis=1\n",
    ")\n",
    "counts_df['fitness_mean_truncated'] = counts_df.apply(\n",
    "    lambda row: np.log((row['actual_count'] + 0.5) / (row['expected_count_mean_truncated'] + 0.5)),\n",
    "    axis=1\n",
    ")\n",
    "counts_df['fitness_median'] = counts_df.apply(\n",
    "    lambda row: np.log((row['actual_count'] + 0.5) / (row['expected_count_median'] + 0.5)),\n",
    "    axis=1\n",
    ")\n",
    "\n",
    "# Read in gene boundaries\n",
    "gene_boundaries_df = pd.read_csv('../results/gene_boundaries.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot distributions of counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Make a plot that summarizes for each mutation type the number of sites being analyzed and the distribution of synonymous mutation counts for that set of sites."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "counts_var = 'truncated_actual_count'\n",
    "print(counts_var)\n",
    "mean_counts_df = counts_df[\n",
    "    counts_df['mut_class'] == 'synonymous'\n",
    "].groupby('mut_type', as_index=False).agg(\n",
    "    mean = (counts_var, 'mean'),\n",
    "    var = (counts_var, 'var'),\n",
    "    std = (counts_var, 'std'),\n",
    "    n = ('mut_type', 'count')\n",
    ")\n",
    "mean_counts_df['var_over_mean'] = mean_counts_df['var'] / mean_counts_df['mean']\n",
    "mean_counts_df['var_over_mean2'] = mean_counts_df['var'] / np.power(mean_counts_df['mean'], 2)\n",
    "mean_counts_df['std_over_mean'] = mean_counts_df['std'] / mean_counts_df['mean']\n",
    "mean_counts_df['mut_type_arrow'] = mean_counts_df['mut_type'].apply(\n",
    "    lambda x: x[0]+r'$\\rightarrow$'+x[1]\n",
    ")\n",
    "mean_counts_df['wt_nt'] = mean_counts_df['mut_type'].apply(lambda x: x[0])\n",
    "mut_order = mean_counts_df['mut_type_arrow'].unique()\n",
    "\n",
    "# Plot data\n",
    "(fig, axs) = plt.subplots(nrows=3, sharex=True, figsize=[5,6])\n",
    "mut_type_pal = ['#1b9e77', '#d95f02', '#7570b3', '#e7298a']\n",
    "\n",
    "# Plot counts distributions\n",
    "data = counts_df[counts_df['mut_class'] == 'synonymous'].copy()\n",
    "sns.boxplot(\n",
    "    x='mut_type_arrow', y=counts_var, data=data, whis=(5,95), showfliers=False,\n",
    "    order=mut_order, ax=axs[0],\n",
    "    hue='wt_nt', dodge=False, palette=mut_type_pal\n",
    ")\n",
    "axs[0].set(\n",
    "    xlabel='', ylabel=f'synonymous mut.\\ncounts per site', yscale='log',\n",
    "    ylim=[0.5, 6e3], yticks=[1, 1e1, 1e2, 1e3]\n",
    ")\n",
    "\n",
    "# Plot variance over mean\n",
    "sns.barplot(\n",
    "    x='mut_type_arrow', y='var_over_mean', data=mean_counts_df,\n",
    "    order=mut_order, ax=axs[1],\n",
    "    hue='wt_nt', dodge=False, palette=mut_type_pal, #color='0.5',\n",
    ")\n",
    "axs[1].axhline(1, ls='--', c='0.25')\n",
    "axs[1].set(\n",
    "    xlabel='mutation type', ylabel=r'$\\frac{{\\sigma^2}}{{\\mu}}$', yscale='log',\n",
    "    ylim=[0.5, 1e3]\n",
    ")\n",
    "axs[1].minorticks_off()\n",
    "axs[1].yaxis.label.set(rotation='horizontal', va='center', ha='right', fontsize=19)\n",
    "\n",
    "# Plot variance over mean squared\n",
    "sns.barplot(\n",
    "    x='mut_type_arrow', y='var_over_mean2', data=mean_counts_df,\n",
    "    order=mut_order, ax=axs[2],\n",
    "    hue='wt_nt', dodge=False, palette=mut_type_pal, #color='0.5',\n",
    ")\n",
    "axs[2].axhline(1, ls='--', c='0.25')\n",
    "axs[2].set(\n",
    "    xlabel='mutation type', ylabel=r'$\\frac{{\\sigma^2}}{{\\mu^2}}$',\n",
    "    yscale='log', ylim=[0.1, 1e1]\n",
    ")\n",
    "axs[2].minorticks_off()\n",
    "axs[2].yaxis.label.set(rotation='horizontal', va='center', ha='right', fontsize=19)\n",
    "axs[2].tick_params('x', labelrotation=90)\n",
    "\n",
    "for i in [0, 1, 2]:\n",
    "    axs[i].grid(axis='y')\n",
    "    axs[i].get_legend().remove()\n",
    "plt.tight_layout()\n",
    "sns.despine()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Show the number of analyzed sites per mutation type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(mean_counts_df[['mut_type', 'n']])\n",
    "# print(\n",
    "#     mean_counts_df[['mut_type_arrow', 'n']]\n",
    "#     .rename(columns={'mut_type_arrow' : 'mutation type', 'n' : 'number of sites'})\n",
    "#     .to_latex(index=False, label='n_sites', formatters={'number of sites' : '{:,.0f}'}, column_format='cc')\n",
    "# )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each mutation type, compare the distribution of counts for synonymous, nonsynonymous, and stop mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mut_class_colors = {\n",
    "    'synonymous' : '#0173b2',\n",
    "    'nonsynonymous' : '#de8f05',\n",
    "    'nonsense' : 'firebrick',\n",
    "    'noncoding' : 'darkviolet',\n",
    "}\n",
    "counts_var = 'truncated_actual_count'\n",
    "order = [\n",
    "    'synonymous',\n",
    "    'nonsynonymous', 'nonsense'\n",
    "]\n",
    "(fig, axs) = plt.subplots(nrows=3, ncols=4, sharex=True, sharey=False, figsize=[10,6])\n",
    "axs = axs.reshape(-1, order='F')\n",
    "\n",
    "data = counts_df[counts_df['gene'].isin(['ORF1a;ORF1ab', 'ORF1ab', 'S', 'E', 'M', 'N'])]\n",
    "\n",
    "for i, (mut_type, mut_data) in enumerate(data.groupby('mut_type_arrow')):\n",
    "\n",
    "    # print(mut_type.replace('$\\\\rightarrow$', ''), mut_data['mut_class'].value_counts())\n",
    "\n",
    "    sns.boxplot(\n",
    "        x='mut_class', y=counts_var, data=mut_data,\n",
    "        order=order, hue='mut_class', hue_order=order, palette=mut_class_colors,\n",
    "        whis=(5,95),# whis=(0,100), # \n",
    "        showfliers=False, ax=axs[i]\n",
    "    )\n",
    "    if sum(mut_data['mut_class'] == 'nonsense') == 0:\n",
    "        axs[i].annotate('n.a.', (0.75, 0.1), xycoords=axs[i].transAxes)\n",
    "    axs[i].set(title=mut_type, xlabel='', ylabel='')\n",
    "    axs[i].tick_params(axis='x', labelrotation=90)\n",
    "\n",
    "fig.text(-0.01, 0.59, 'mutation counts', va='center', rotation='vertical')\n",
    "plt.suptitle('data from genes under strong purifying selection')\n",
    "plt.tight_layout()\n",
    "sns.despine()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Across sites where a given pair of mutation types both result in synonymous mutations, is there a correlation in counts between mutation types?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def corrfunc(x, y, ax=None, **kws):\n",
    "    \"\"\"Plot the correlation coefficient in the top left hand corner of a plot.\"\"\"\n",
    "    r, _ = scipy.stats.pearsonr(x, y)\n",
    "    r2 = math.pow(r, 2)\n",
    "    ax = ax or plt.gca()\n",
    "    ax.annotate('$R^2$' + f' = {r2:.2f}', xy=(.05, .875), xycoords=ax.transAxes)\n",
    "\n",
    "counts_var = 'truncated_actual_count'\n",
    "data = counts_df[\n",
    "    (counts_df['mut_class'] == 'synonymous')\n",
    "].pivot(index='site', columns='mut_type', values=counts_var)\n",
    "mut_sets = [\n",
    "    ['AC', 'AG', 'AT'],\n",
    "    ['CA', 'CG', 'CT'],\n",
    "    ['GA', 'GC', 'GT'],\n",
    "    ['TA', 'TC', 'TG']\n",
    "]\n",
    "for mut_set in mut_sets:\n",
    "    g = sns.pairplot(\n",
    "        data[mut_set].dropna().rename(columns={x:x[0]+r'$\\rightarrow$'+x[1] for x in mut_set}),\n",
    "        corner=True, kind='hist',\n",
    "        plot_kws={'bins':15},\n",
    "        height=2\n",
    "    )\n",
    "    g.map_lower(corrfunc)\n",
    "    for ax in g.diag_axes:\n",
    "        ax.set_visible(False)\n",
    "    for i, y_var in enumerate(g.y_vars):\n",
    "        for j, x_var in enumerate(g.x_vars):\n",
    "            if x_var == y_var:\n",
    "                g.axes[i, j].set_visible(False)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot the distribution of counts as a function of primary sequence, using a sliding window to smooth the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mut_types = sorted(list(counts_df['mut_type'].unique()))\n",
    "counts_var = 'truncated_actual_count'\n",
    "print('counts_var', counts_var)\n",
    "window_flank_size = 1000\n",
    "print('window size', window_flank_size*2+1)\n",
    "plot_poisson = False\n",
    "rolling_dfs = []\n",
    "means_dict = defaultdict(list)\n",
    "window_stats_dict = defaultdict(list)\n",
    "for (mut_type, mut_type_data) in counts_df.groupby('mut_type'):\n",
    "\n",
    "    print(mut_type)\n",
    "\n",
    "    # Plot data for syn, nonsyn, stop muts\n",
    "    (fig, axs) = plt.subplots(\n",
    "        nrows=3, ncols=2, figsize=[7, 5], sharex='col', sharey='row',\n",
    "        height_ratios=[1,3,3], width_ratios=[7,1]\n",
    "    )\n",
    "    axs = axs.reshape(-1)\n",
    "    max_window_median = 0\n",
    "    for (i, (mut_class, data)) in enumerate(mut_type_data.groupby('mut_class')):\n",
    "\n",
    "        if mut_class in ['noncoding']:\n",
    "            continue\n",
    "\n",
    "        # Compute sliding-window medians\n",
    "        assert len(data['site']) == len(set(data['site']))\n",
    "        window_dict = defaultdict(list)\n",
    "        for site in data['site']:\n",
    "            data_i = data[data['site'].between(site-window_flank_size, site+window_flank_size)]\n",
    "            window_dict['site'].append(site)\n",
    "            if len(data_i) == 0:\n",
    "                window_dict[counts_var].append(np.nan)\n",
    "            else:\n",
    "                window_dict[counts_var].append(data_i[counts_var].median())\n",
    "        rolling_data = pd.DataFrame(window_dict)\n",
    "        rolling_data['mut_type'] = mut_type\n",
    "        rolling_data['mut_class'] = mut_class\n",
    "        rolling_dfs.append(rolling_data)\n",
    "        if rolling_data[counts_var].max() > max_window_median:\n",
    "            max_window_median = rolling_data[counts_var].max()\n",
    "        \n",
    "        # Plot sliding-window medians and scatter plot\n",
    "        sns.lineplot(\n",
    "            x='site', y=counts_var, data=rolling_data, ax=axs[2],\n",
    "            color=mut_class_colors[mut_class], lw=2\n",
    "        )\n",
    "        if mut_class == 'stop':\n",
    "            zorder = 10\n",
    "            alpha = 1\n",
    "        else:\n",
    "            zorder = 1\n",
    "            alpha = 0.5\n",
    "        if mut_class in ['synonymous']:\n",
    "            sns.scatterplot(\n",
    "                x='site', y=counts_var, data=data, ax=axs[4],\n",
    "                alpha=alpha, color=mut_class_colors[mut_class],\n",
    "                zorder=zorder\n",
    "            )\n",
    "\n",
    "    # Plot vertical lines at indicated sites\n",
    "    coords = []\n",
    "    if mut_type in ['AT', 'CG', 'GC']:\n",
    "        coords += [21555]\n",
    "    elif mut_type in ['CT']:\n",
    "        coords += [13483]\n",
    "    for coord in coords:\n",
    "        for i in [2, 4]:\n",
    "            axs[i].axvline(coord, ls='--', c='0.25')\n",
    "\n",
    "    axs[4].set(ylabel='synonymous\\nmutation counts')\n",
    "    axs[2].set(\n",
    "        ylabel='sliding-window\\nmedian (2 kb)',\n",
    "    )\n",
    "    axs[2].grid()\n",
    "    sns.despine()\n",
    "\n",
    "    # Add gene annotation plot\n",
    "    for (i, row) in gene_boundaries_df.iterrows():\n",
    "        start = row['min_site']\n",
    "        end = row['max_site']\n",
    "        gene = row['gene']\n",
    "        if gene in ['ORF1ab']:\n",
    "            continue\n",
    "        elif gene in ['ORF1a', 'ORF1b', 'S', 'E', 'M', 'N']:\n",
    "            y_min = -0.5\n",
    "            annotate = True\n",
    "        else:\n",
    "            y_min = 0.5\n",
    "            annotate = False\n",
    "        rectangle = matplotlib.patches.Rectangle((start, y_min), end-start, 0.75, edgecolor='0.25', facecolor='white')\n",
    "        axs[0].add_patch(rectangle)\n",
    "        if annotate:\n",
    "            rx, ry = rectangle.get_xy()\n",
    "            cx = rx + rectangle.get_width()/2.0\n",
    "            cy = ry - 0.05 + rectangle.get_height()/2.0\n",
    "            if gene in ['E']:\n",
    "                cx -= 450\n",
    "            axs[0].annotate(\n",
    "                gene, (cx, cy), color='black', ha='center', va='center',\n",
    "                fontsize=9\n",
    "            )\n",
    "    axs[0].set(ylim=[-0.55,1.25], yticks=[])\n",
    "    sns.despine(left=True, bottom=True, ax=axs[0])\n",
    "    axs[0].xaxis.set_tick_params(which='both', bottom=False, labelbottom=False, labeltop=False)\n",
    "\n",
    "    # Plot the histogram of counts\n",
    "    data = mut_type_data[mut_type_data['mut_class'] == 'synonymous'].copy()\n",
    "    nbins = 20\n",
    "    min_count = data[counts_var].min()\n",
    "    max_count = data[counts_var].max()\n",
    "    bin_size = (max_count-min_count)/nbins\n",
    "    bins = np.arange(min_count, max_count+bin_size, bin_size)\n",
    "    sns.histplot(\n",
    "        y=counts_var, data=data, bins=bins, ax=axs[5],\n",
    "        color=mut_class_colors['synonymous'],\n",
    "    )\n",
    "\n",
    "    # Plot histogram of poisson distribution with same mean\n",
    "    if plot_poisson:\n",
    "        mean = data[counts_var].mean()\n",
    "        xs = np.arange(0, data[counts_var].max()+1, 1)\n",
    "        poisson = scipy.stats.poisson(mean)\n",
    "        ys = poisson.pmf(xs) * len(data)\n",
    "        sim_counts = poisson.rvs(size=len(data), random_state=1)\n",
    "        sns.histplot(y=sim_counts, bins=bins, color='0.5', ax=axs[5])\n",
    "        axs[5].axhline(mean, ls='--', c='0.25')\n",
    "    sns.despine(bottom=True, ax=axs[5])\n",
    "    axs[5].set(xlabel='', xticks=[])\n",
    "\n",
    "    axs[1].axis('off')\n",
    "    axs[3].axis('off')\n",
    "    axs[0].set(xlim=[-200,30200])\n",
    "    axs[4].grid()\n",
    "\n",
    "    plt.suptitle(mut_type[0]+r'$\\rightarrow$'+mut_type[1], x=0.46, y=0.95)\n",
    "    plt.tight_layout(w_pad=0)\n",
    "    plt.show()\n",
    "\n",
    "rolling_df = pd.concat(rolling_dfs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot simplified versions of the above plots for just synonymous mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mut_types = ['AG', 'AT', 'CT']\n",
    "counts_var = 'truncated_actual_count'\n",
    "print('counts_var', counts_var)\n",
    "window_flank_size = 1000\n",
    "print('window size', window_flank_size*2+1)\n",
    "\n",
    "for mut_type in mut_types:\n",
    "\n",
    "    # Get data\n",
    "    print(mut_type)\n",
    "    data = counts_df[\n",
    "        (counts_df['mut_type'] == mut_type) &\n",
    "        (counts_df['mut_class'] == 'synonymous')\n",
    "    ].copy()\n",
    "    rolling_data = rolling_df[\n",
    "        (rolling_df['mut_type'] == mut_type) &\n",
    "        (rolling_df['mut_class'] == 'synonymous')\n",
    "    ]\n",
    "\n",
    "    # Plot data\n",
    "    (fig, axs) = plt.subplots(\n",
    "        nrows=2, ncols=2, figsize=[7, 3.5], sharex='col', sharey='row',\n",
    "        height_ratios=[1,3.5], width_ratios=[7,1]\n",
    "    )\n",
    "    axs = axs.reshape(-1)\n",
    "    \n",
    "    # Plot sliding-window medians and scatter plot\n",
    "    sns.lineplot(\n",
    "        x='site', y=counts_var, data=rolling_data, ax=axs[2],\n",
    "        color='0.25', lw=2\n",
    "    )\n",
    "    sns.scatterplot(\n",
    "        x='site', y=counts_var, data=data, ax=axs[2],\n",
    "        alpha=0.5, color=mut_class_colors['synonymous'],\n",
    "        zorder=1\n",
    "    )\n",
    "\n",
    "    # Plot vertical lines at indicated sites\n",
    "    coords = []\n",
    "    if mut_type in ['AT', 'CG', 'GC']:\n",
    "        coords += [21555]\n",
    "    elif mut_type in ['CT']:\n",
    "        coords += [13483]\n",
    "    for coord in coords:\n",
    "        axs[2].axvline(coord, ls='--', c='0.25')\n",
    "\n",
    "    axs[2].set(ylabel='synonymous\\nmutation counts')\n",
    "    axs[2].grid()\n",
    "    sns.despine()\n",
    "\n",
    "    # Add gene annotation plot\n",
    "    for (i, row) in gene_boundaries_df.iterrows():\n",
    "        start = row['min_site']\n",
    "        end = row['max_site']\n",
    "        gene = row['gene']\n",
    "        if gene in ['ORF1ab']:\n",
    "            continue\n",
    "        elif gene in ['ORF1a', 'ORF1b', 'S', 'E', 'M', 'N']:\n",
    "            y_min = -0.5\n",
    "            annotate = True\n",
    "        else:\n",
    "            y_min = 0.5\n",
    "            annotate = False\n",
    "        rectangle = matplotlib.patches.Rectangle((start, y_min), end-start, 0.75, edgecolor='0.25', facecolor='white')\n",
    "        axs[0].add_patch(rectangle)\n",
    "        if annotate:\n",
    "            rx, ry = rectangle.get_xy()\n",
    "            cx = rx + rectangle.get_width()/2.0\n",
    "            cy = ry - 0.05 + rectangle.get_height()/2.0\n",
    "            if gene in ['E']:\n",
    "                cx -= 450\n",
    "            axs[0].annotate(\n",
    "                gene, (cx, cy), color='black', ha='center', va='center',\n",
    "                fontsize=9\n",
    "            )\n",
    "    axs[0].set(ylim=[-0.55,1.25], yticks=[])\n",
    "    sns.despine(left=True, bottom=True, ax=axs[0])\n",
    "    axs[0].xaxis.set_tick_params(which='both', bottom=False, labelbottom=False, labeltop=False)\n",
    "\n",
    "    # Plot the histogram of counts\n",
    "    nbins = 20\n",
    "    min_count = data[counts_var].min()\n",
    "    max_count = data[counts_var].max()\n",
    "    bin_size = (max_count-min_count)/nbins\n",
    "    bins = np.arange(min_count, max_count+bin_size, bin_size)\n",
    "    sns.histplot(\n",
    "        y=counts_var, data=data, bins=bins, ax=axs[3],\n",
    "        color=mut_class_colors['synonymous'],\n",
    "    )\n",
    "    sns.despine(bottom=True, ax=axs[3])\n",
    "    axs[3].set(xlabel='', xticks=[])\n",
    "\n",
    "    axs[1].axis('off')\n",
    "    axs[0].set(xlim=[-200,30200])\n",
    "\n",
    "    plt.suptitle(mut_type[0]+r'$\\rightarrow$'+mut_type[1], x=0.46, y=0.9)\n",
    "    plt.tight_layout(w_pad=0)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Make a summary plot showing the sliding-window medians of synonymous mutations for each mutation type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.patches as patches"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the median of each mutation type across all sites\n",
    "#counts_var = 'truncated_actual_count'\n",
    "print(counts_var)\n",
    "median_data = counts_df[counts_df['mut_class'] == 'synonymous'].groupby('mut_type', as_index=False).agg(\n",
    "    median = (counts_var, 'median')\n",
    ")\n",
    "\n",
    "# Get window data for synonymous mutations and then merge with the dataframe\n",
    "# giving global averages\n",
    "data = rolling_df[rolling_df['mut_class'] == 'synonymous'].copy()\n",
    "data.rename(columns={\n",
    "    counts_var : 'observed data',\n",
    "}, inplace=True)\n",
    "data = data.melt(\n",
    "    id_vars=['site', 'mut_type'], value_vars=['observed data'],\n",
    "    value_name='counts', var_name='count_type'\n",
    ").merge(median_data[['mut_type', 'median']], on='mut_type', how='left')\n",
    "data['fold_change_from_median'] = data['counts'] / data['median']\n",
    "\n",
    "var_data = data.groupby('mut_type', as_index=False).agg(\n",
    "    var = ('fold_change_from_median', 'var')\n",
    ")\n",
    "data = data.merge(var_data, on='mut_type')\n",
    "data.sort_values('var', ascending=False, inplace=True)\n",
    "\n",
    "data['mut_type_arrow'] = data['mut_type'].apply(\n",
    "    lambda x: x[0]+r'$\\rightarrow$'+x[1]\n",
    ")\n",
    "\n",
    "data = data[data['count_type'] == 'observed data']\n",
    "\n",
    "# Plot data\n",
    "min_site = data['site'].min() - 100\n",
    "max_site = data['site'].max() + 100\n",
    "(fig, axs) = plt.subplots(nrows=12, sharex=True, sharey=True, figsize=[4,6])\n",
    "axs[0].scatter([13483, 21555], [2.6, 2.6], c='0.3', marker='v')\n",
    "for (i, (mut_type, mut_type_data)) in enumerate(data.groupby('mut_type_arrow')):\n",
    "    sns.lineplot(\n",
    "        x='site', y='fold_change_from_median', data=mut_type_data,\n",
    "        ax=axs[i], color=mut_class_colors['synonymous']\n",
    "    )\n",
    "    rect = patches.Rectangle(\n",
    "        (min_site, 0.5), max_site - min_site, 1.5,\n",
    "        linewidth=0, color='0.92', zorder=0\n",
    "    )\n",
    "    axs[i].add_patch(rect)\n",
    "    axs[i].plot([min_site, max_site], [1, 1], c='k', lw=0.5, zorder=0)\n",
    "    axs[i].set_yscale('log', base=2)\n",
    "    axs[i].set(ylabel='', yticks=[], xticks=[]) # 0.5, 1, 4\n",
    "    axs[i].set_title(mut_type, loc='left', y=0.11, x=-0.1)\n",
    "    sns.despine(bottom=True, left=True, ax=axs[i])\n",
    "\n",
    "plt.tight_layout(h_pad=0)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot distribution of counts for paired vs. unpaired bases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "classification_var = 'ss_prediction' # 'biphasic_at' # \n",
    "counts_var = 'truncated_actual_count'\n",
    "print(classification_var, counts_var)\n",
    "if classification_var == 'ss_prediction':\n",
    "    hue_order = ['paired', 'unpaired']\n",
    "elif classification_var == 'biphasic_at':\n",
    "    hue_order = ['phase_1', 'phase_2']\n",
    "else:\n",
    "    raise ValueError()\n",
    "(fig, axs) = plt.subplots(nrows=3, ncols=4, sharex=True, sharey=False, figsize=[10,6])\n",
    "axs = axs.reshape(-1, order='F')\n",
    "order = ['synonymous', 'nonsynonymous', 'nonsense']\n",
    "for i, (mut_type, mut_data) in enumerate(counts_df.groupby('mut_type_arrow')):\n",
    "    sns.boxplot(\n",
    "        x='mut_class', y=counts_var, data=mut_data,\n",
    "        hue=classification_var, hue_order=hue_order, palette=['0.5', '0.75'],\n",
    "        whis=(5,95), showfliers=False, ax=axs[i], order=order\n",
    "    )\n",
    "    if sum(mut_data['mut_class'] == 'nonsense') == 0:\n",
    "        axs[i].annotate('n.a.', (0.75, 0.1), xycoords=axs[i].transAxes)\n",
    "    axs[i].set(title=mut_type, xlabel='', ylabel='')\n",
    "    axs[i].tick_params(axis='x', labelrotation=90)\n",
    "    axs[i].get_legend().remove()\n",
    "\n",
    "fig.text(-0.01, 0.59, 'mutation counts', va='center', rotation='vertical')\n",
    "plt.tight_layout()\n",
    "sns.despine()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For each mutation type, plot the difference in median synonymous counts for paired vs. unpaired"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute fold change in median counts between unpaired vs. paired sites or sites before\n",
    "# or after the biphasic change at site 21,555\n",
    "classification_var = 'ss_paired' # 'nt_site_before_21555' # \n",
    "counts_var = 'truncated_actual_count'\n",
    "ratio_label_dict = {\n",
    "    'ss_paired' : '(unpaired : paired)',\n",
    "    'nt_site_before_21555' : '(post- : pre- site 21,555)'\n",
    "}\n",
    "data = (\n",
    "    counts_df[counts_df['mut_class'] == 'synonymous']\n",
    "    .groupby(['mut_type', classification_var])[[counts_var]].median()\n",
    "    .reset_index()\n",
    "    .pivot_table(index='mut_type', columns=classification_var, values=counts_var)\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "data['fold_diff'] = data[False] / data[True]\n",
    "data['wt_nt'] = data['mut_type'].apply(lambda x: x[0])\n",
    "data['mut_type_arrow'] = data['mut_type'].apply(\n",
    "    lambda x: x[0]+r'$\\rightarrow$'+x[1]\n",
    ")\n",
    "fold_change_data = data.copy()\n",
    "\n",
    "# Plot data\n",
    "mut_type_order = [\n",
    "    'AC', 'AG', 'AT',\n",
    "    'CA', 'CG', 'CT',\n",
    "    'GA', 'GC', 'GT',\n",
    "    'TA', 'TC', 'TG'\n",
    "]\n",
    "mut_type_arrow_order = [x[0]+r'$\\rightarrow$'+x[1] for x in mut_type_order]\n",
    "plt.figure(figsize=[5,3])\n",
    "sns.barplot(\n",
    "    x='mut_type_arrow', y='fold_diff', data=fold_change_data,\n",
    "    hue='wt_nt', dodge=False, palette=mut_type_pal\n",
    ")\n",
    "\n",
    "# Use randomization to test the null hypothesis that the medians are >1 due\n",
    "# to random chance\n",
    "dfs = []\n",
    "rs = np.random.RandomState(seed=1)\n",
    "for i in range(10):\n",
    "    \n",
    "    # Get data for synonymous mutations\n",
    "    data = counts_df[counts_df['mut_class'] == 'synonymous'].copy()\n",
    "    \n",
    "    # Get data for each mutation type, shuffle counts between sites, and then\n",
    "    # concatenate the data across mutation teyps\n",
    "    shuffled_dfs = []\n",
    "    for (mut_type, data_i) in data.groupby('mut_type'):\n",
    "        data_i['shuffled_counts'] = list(data_i[counts_var].sample(frac=1, random_state=rs))\n",
    "        shuffled_dfs.append(data_i)\n",
    "    data = pd.concat(shuffled_dfs)\n",
    "\n",
    "    # Compute the fold difference between medians using shuffled counts\n",
    "    data = (\n",
    "        data\n",
    "        .groupby(['mut_type', classification_var])[['shuffled_counts']].median()\n",
    "        .reset_index()\n",
    "        .pivot_table(index='mut_type', columns=classification_var, values='shuffled_counts')\n",
    "        .reset_index()\n",
    "    )\n",
    "    data['fold_diff'] = data[False] / data[True]\n",
    "    data['rand_trial'] = i\n",
    "    dfs.append(data)\n",
    "\n",
    "data = pd.concat(dfs)\n",
    "rand_data = data.pivot(index='rand_trial', columns='mut_type', values='fold_diff')\n",
    "\n",
    "# Compute p values from randomized data and add asterisk to plot if\n",
    "# a mutation type's corrected p value is < 0.05\n",
    "print('mut_type', 'fold_diff', 'p_val')\n",
    "for (i, row) in fold_change_data.iterrows():\n",
    "    mut_type = row['mut_type']\n",
    "    frac_rand_greater = sum(rand_data[mut_type] >= row['fold_diff']) / len(rand_data)\n",
    "    print(mut_type, round(row['fold_diff'], 3), round(frac_rand_greater, 3))\n",
    "    if frac_rand_greater * 12 < 0.05:\n",
    "        y = row['fold_diff'] + 0.1\n",
    "        plt.annotate('*', (i,y), weight='bold', va='center', ha='center')\n",
    "\n",
    "plt.tick_params('x', labelrotation=90)\n",
    "ax = plt.gca()\n",
    "ax.set(\n",
    "    xlabel='mutation type',\n",
    "    ylabel=f'ratio of median syn.\\nmutataion counts\\n{ratio_label_dict[classification_var]}'\n",
    ")\n",
    "ax.grid(axis='y')\n",
    "ax.get_legend().remove()\n",
    "plt.tight_layout()\n",
    "sns.despine()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot a sliding-window average of the fraction of nucleotides that are paired in the genome's predicted secondary structure."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute average ss content over sliding window\n",
    "window_dict = defaultdict(list)\n",
    "window_flank_size = 1000\n",
    "data = counts_df[\n",
    "    counts_df['mut_class'] == 'synonymous'\n",
    "].drop_duplicates(subset='site')\n",
    "dfs = []\n",
    "for site in data['site']:\n",
    "    data_i = data[data['site'].between(site-window_flank_size, site+window_flank_size)]\n",
    "    data_i = (\n",
    "        data_i\n",
    "        .groupby('wt_nt', as_index=False)[['ss_prediction']]\n",
    "        .value_counts()\n",
    "        .pivot(index='wt_nt', columns='ss_prediction', values='count')\n",
    "        .reset_index()\n",
    "    )\n",
    "    data_i['site'] = site\n",
    "    data_i['frac_ss'] = data_i['paired'] / (data_i['paired'] + data_i['unpaired'])\n",
    "    dfs.append(data_i)\n",
    "ss_window_df = pd.concat(dfs)\n",
    "ss_window_df\n",
    "\n",
    "# Plot sliding-window average\n",
    "(fig, axs) = plt.subplots(\n",
    "    nrows=2, figsize=[6, 3], sharex=True,\n",
    "    height_ratios=[1,3]\n",
    ")\n",
    "sns.lineplot(\n",
    "    x='site', y='frac_ss', data=ss_window_df,\n",
    "    hue='wt_nt', palette=mut_type_pal, ax=axs[1]\n",
    ")\n",
    "axs[1].legend(bbox_to_anchor=(1,1))\n",
    "axs[1].set(ylabel='fraction of paired sites\\n(sliding-window average)')\n",
    "\n",
    "# Add gene annotation plot\n",
    "for (i, row) in gene_boundaries_df.iterrows():\n",
    "    start = row['min_site']\n",
    "    end = row['max_site']\n",
    "    gene = row['gene']\n",
    "    if gene in ['ORF1ab']:\n",
    "        continue\n",
    "    elif gene in ['ORF1a', 'ORF1b', 'S', 'E', 'M', 'N']:\n",
    "        y_min = -0.5\n",
    "        annotate = True\n",
    "    else:\n",
    "        y_min = 0.5\n",
    "        annotate = False\n",
    "    rectangle = matplotlib.patches.Rectangle((start, y_min), end-start, 0.75, edgecolor='0.25', facecolor='white')\n",
    "    axs[0].add_patch(rectangle)\n",
    "    if annotate:\n",
    "        rx, ry = rectangle.get_xy()\n",
    "        cx = rx + rectangle.get_width()/2.0\n",
    "        cy = ry - 0.05 + rectangle.get_height()/2.0\n",
    "        if gene in ['E', 'M']:\n",
    "            cy -= 0.7\n",
    "        axs[0].annotate(\n",
    "            gene, (cx, cy), color='black', ha='center', va='center',\n",
    "            fontsize=9\n",
    "        )\n",
    "\n",
    "sns.despine()\n",
    "axs[0].set(ylim=[-0.55,1.25], yticks=[])\n",
    "sns.despine(left=True, bottom=True, ax=axs[0])\n",
    "axs[0].xaxis.set_tick_params(which='both', bottom=False, labelbottom=False, labeltop=False)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Examine the relationship between counts and local sequence context"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, for each mutation type, group sites by 3-mer motifs and compute median counts per motif, only doing so for motifs with at least 10 sites with data for that mutation type."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compute global medians\n",
    "counts_var = 'truncated_actual_count'\n",
    "global_medians = (\n",
    "    counts_df[counts_df['mut_class'] == 'synonymous']\n",
    "    .groupby(['mut_type'], as_index=False)\n",
    "    .agg(global_median_counts=pd.NamedAgg(column=counts_var, aggfunc='median'))\n",
    ")\n",
    "\n",
    "# Compute motif medians for motifs with at least 10 sites, then merge with\n",
    "# data on global medians\n",
    "motif_medians = (\n",
    "    counts_df[counts_df['mut_class'] == 'synonymous']\n",
    "    .groupby(['mut_type', 'motif'], as_index=False)\n",
    "    .agg(\n",
    "        median_counts=pd.NamedAgg(column=counts_var, aggfunc='median'),\n",
    "        median_counts_pre_omicron=pd.NamedAgg(column='truncated_actual_count_pre_omicron', aggfunc='median'),\n",
    "        median_counts_omicron=pd.NamedAgg(column='truncated_actual_count_omicron', aggfunc='median'),\n",
    "        n=pd.NamedAgg(column='motif', aggfunc='count'),\n",
    "    )\n",
    "    .query('n >= 10')\n",
    "    .merge(global_medians, on='mut_type')\n",
    ")\n",
    "motif_medians['fold_change_from_global'] = motif_medians['median_counts'] / motif_medians['global_median_counts']\n",
    "\n",
    "# Add a column giving a reference motif that the motif for mutation types\n",
    "# away from T or C, but the reverse-complement motif for mutation types\n",
    "# away from A or G, allowing for comparison of reverse-complement mutation\n",
    "# types\n",
    "def get_ref_motif(mut_type, motif):\n",
    "    if mut_type[0] in ['A', 'G']:\n",
    "        return str(Seq(motif).reverse_complement())\n",
    "    else:\n",
    "        return motif\n",
    "motif_medians['ref_motif'] = motif_medians.apply(\n",
    "    lambda row: get_ref_motif(row['mut_type'], row['motif']), axis=1\n",
    ")\n",
    "\n",
    "motif_medians['mut_type_arrow'] = motif_medians['mut_type'].apply(\n",
    "    lambda x: x[0]+r'$\\rightarrow$'+x[1]\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, for each mutation type, plot the distribution of mutation counts as a function of a site's 3-mer motif."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "counts_var = 'truncated_actual_count'\n",
    "mut_classes = [\n",
    "    'synonymous',\n",
    "    # 'nonsynonymous',\n",
    "    # 'stop'\n",
    "]\n",
    "mut_types = [\n",
    "    'TG', 'AC',\n",
    "    'TA', 'AT', \n",
    "    'TC', 'AG',\n",
    "    'CT', 'GA',\n",
    "    'CA', 'GT',\n",
    "    'CG', 'GC',\n",
    "]\n",
    "for mut_type in mut_types:\n",
    "\n",
    "    # Get data for mutation type\n",
    "    print(mut_type)\n",
    "    data = counts_df[\n",
    "        (counts_df['mut_class'].isin(mut_classes)) &\n",
    "        (counts_df['mut_type'] == mut_type)\n",
    "    ].copy()\n",
    "    global_median = data[data['mut_class'] == 'synonymous'][counts_var].median()\n",
    "    motif_medians_data = motif_medians[motif_medians['mut_type'] == mut_type]\n",
    "\n",
    "    # Report stats\n",
    "    print('global median', global_median)\n",
    "    print('min median', round(motif_medians_data['median_counts'].min()))\n",
    "    print('max median', round(motif_medians_data['median_counts'].max()))\n",
    "    print('max/min median', round(motif_medians_data['median_counts'].max()/motif_medians_data['median_counts'].min(), 2))\n",
    "\n",
    "    # Sort motifs by their meian, then plot the distribution of counts for each motif\n",
    "    data = data.merge(motif_medians_data, on=['mut_type', 'motif'], how='right')\n",
    "    data.sort_values('median_counts', inplace=True)\n",
    "    if len(mut_classes) == 1:\n",
    "        plt.figure(figsize=[8,3])\n",
    "    else:\n",
    "        plt.figure(figsize=[12,3])\n",
    "    sns.boxplot(\n",
    "        x='motif', y=counts_var, data=data,\n",
    "        hue='mut_class', hue_order=mut_classes, palette=mut_class_colors,\n",
    "        #showmeans=True,\n",
    "        showfliers=False, whis=0,\n",
    "        linewidth=0,\n",
    "        medianprops={\"linewidth\": 2},\n",
    "        boxprops={'alpha':.3},\n",
    "        zorder=10,\n",
    "    )\n",
    "    sns.stripplot(\n",
    "        x='motif', y=counts_var, data=data,\n",
    "        hue='mut_class', hue_order=mut_classes, palette=mut_class_colors,\n",
    "        dodge=True, alpha=0.5, s=7\n",
    "    )\n",
    "    plt.axhline(global_median, c='0.75', ls='--')\n",
    "    plt.legend().remove()\n",
    "    sns.despine()\n",
    "    plt.title(mut_type[0]+r'$\\rightarrow$'+mut_type[1])\n",
    "    plt.ylabel('synonymous mutation counts')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot counts per motif for a specific mutation type and subset of sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mut_type = 'CT'\n",
    "title_suffix = '(paired)'\n",
    "data = counts_df[\n",
    "    (counts_df['mut_class'].isin(mut_classes)) &\n",
    "    (counts_df['ss_prediction'] == 'paired') &\n",
    "    (counts_df['mut_type'] == mut_type)\n",
    "].copy()\n",
    "\n",
    "# Compute the mean counts of syn mutations for each motif\n",
    "print(mut_type)\n",
    "global_median = data[data['mut_class'] == 'synonymous'][counts_var].median()\n",
    "medians_df = (\n",
    "    data[data['mut_class'] == 'synonymous']\n",
    "    .groupby(['mut_type', 'motif'], as_index=False)\n",
    "    .agg(\n",
    "        median_counts=pd.NamedAgg(column=counts_var, aggfunc='median'),\n",
    "        n=pd.NamedAgg(column='motif', aggfunc='count'),\n",
    "    )\n",
    "    .query('n >= 10')\n",
    "    .merge(global_medians, on='mut_type')\n",
    ")\n",
    "\n",
    "# Report stats\n",
    "print('global median', global_median)\n",
    "print('min median', round(medians_df['median_counts'].min()))\n",
    "print('max median', round(medians_df['median_counts'].max()))\n",
    "print('max/min median', round(medians_df['median_counts'].max()/medians_df['median_counts'].min(), 2))\n",
    "\n",
    "# Sort motifs by their mean, then plot the distribution of counts for each motif\n",
    "data = data.merge(medians_df, on=['mut_type', 'motif'])\n",
    "data.sort_values('median_counts', inplace=True)\n",
    "\n",
    "if len(mut_classes) == 1:\n",
    "    plt.figure(figsize=[8,3])\n",
    "else:\n",
    "    plt.figure(figsize=[12,3])\n",
    "sns.boxplot(\n",
    "    x='motif', y=counts_var, data=data,\n",
    "    hue='mut_class', hue_order=mut_classes, palette=mut_class_colors,\n",
    "    #showmeans=True,\n",
    "    showfliers=False, whis=0,\n",
    "    linewidth=0,\n",
    "    medianprops={\"linewidth\": 2},\n",
    "    boxprops={'alpha':.3},\n",
    "    zorder=10,\n",
    ")\n",
    "sns.stripplot(\n",
    "    x='motif', y=counts_var, data=data,\n",
    "    hue='mut_class', hue_order=mut_classes, palette=mut_class_colors,\n",
    "    dodge=True, alpha=0.5, s=7\n",
    ")\n",
    "plt.axhline(global_median, c='0.75', ls='--')\n",
    "plt.legend().remove()\n",
    "sns.despine()\n",
    "plt.title(mut_type[0]+r'$\\rightarrow$'+mut_type[1]+f' {title_suffix}')\n",
    "plt.ylabel('synonymous mutation counts')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot the correlation in medians between reverse-complement motifs and mutation types."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mut_type_pairs = [\n",
    "    ['TG', 'AC'],\n",
    "    ['TA', 'AT'],\n",
    "    ['TC', 'AG'],\n",
    "    ['CT', 'GA'],\n",
    "    ['CA', 'GT'],\n",
    "    ['CG', 'GC'],\n",
    "]\n",
    "mirror_metadata = {}\n",
    "for mut_type_pair in mut_type_pairs:\n",
    "    data = (\n",
    "        motif_medians[motif_medians['mut_type'].isin(mut_type_pair)]\n",
    "        .pivot(index='ref_motif', columns='mut_type', values='median_counts')\n",
    "        .dropna()\n",
    "    )\n",
    "    plt.figure(figsize=[2.5, 2.5])\n",
    "    x = mut_type_pair[1]\n",
    "    y = mut_type_pair[0]\n",
    "    sns.scatterplot(x=x, y=y, data=data, s=85, alpha=1.0)\n",
    "    r = data[x].corr(data[y])\n",
    "    (m, b) = np.polyfit(data[x], data[y], 1)\n",
    "    plt.annotate(f'R = {round(r, 2)}', [0.05, 0.9], xycoords='axes fraction')\n",
    "    mirror_metadata[mut_type_pair[0]] = [round(r, 4), round(m, 2)]\n",
    "    print(mut_type_pair)\n",
    "    print('r =', round(r, 3),)\n",
    "    print('m =', round(m, 3), 'b =', round(b, 3))\n",
    "    min_val = data.min().min()\n",
    "    max_val = data.max().max()\n",
    "    if mut_type_pair[0][0] == 'T':\n",
    "        plt.plot([min_val, max_val], [min_val, max_val], ls='--', c='0.25')\n",
    "    plt.xlabel(mut_type_pair[1][0]+r'$\\rightarrow$'+mut_type_pair[1][1])\n",
    "    plt.ylabel(mut_type_pair[0][0]+r'$\\rightarrow$'+mut_type_pair[0][1])\n",
    "    sns.despine()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Plot the distribution of motif medians relative to the global median"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sort the input data by pairs of mutation types\n",
    "mut_types_order = []\n",
    "for mut_type_pair in mut_type_pairs:\n",
    "    mut_types_order += mut_type_pair\n",
    "mut_type_indices = {mut_type : i for (i, mut_type) in enumerate(mut_types_order)}\n",
    "motif_medians['mut_type_index'] = motif_medians['mut_type'].apply(lambda x: mut_type_indices[x])\n",
    "motif_medians.sort_values('mut_type_index', inplace=True)\n",
    "\n",
    "# Plot data\n",
    "plt.figure(figsize=[12,3])\n",
    "sns.stripplot(x='mut_type_arrow', y='fold_change_from_global', data=motif_medians, s=10, alpha=0.5)\n",
    "plt.axhline(1, ls='--', c='0.25')\n",
    "\n",
    "for (i, mut_type) in enumerate(mut_types_order):\n",
    "    if mut_type in mirror_metadata:\n",
    "        (r, m) = mirror_metadata[mut_type]\n",
    "        #r = round(r, 2)\n",
    "        r = f'R = {round(r, 2)}'\n",
    "        # if p < 0.001:\n",
    "        #     p = 'p < 0.001'\n",
    "        # else:\n",
    "        #     p = f'p = {round(p, 3)}'\n",
    "        plt.plot([i,i+1], [8,8], c='k')\n",
    "        plt.annotate(f'{r}', [i+0.5, 9.5], ha='center', va='bottom')\n",
    "\n",
    "plt.tick_params('x', labelrotation=90)\n",
    "plt.xlabel('mutation type')\n",
    "plt.ylabel('motif median / global median')\n",
    "ax = plt.gca()\n",
    "ax.set_yscale('log', base=2)\n",
    "plt.yticks([1/16, 1/8, 1/4, 1/2, 1, 2, 4, 8])\n",
    "plt.grid(axis='y')\n",
    "plt.minorticks_off()\n",
    "sns.despine()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Analyze how much 3-mer motif medians differ between paired vs. unpaired sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get data\n",
    "counts_var = 'truncated_actual_count'\n",
    "data = (\n",
    "    counts_df[counts_df['mut_class'] == 'synonymous']\n",
    "    .groupby(['mut_type_arrow', 'motif', 'ss_prediction'], as_index=False)\n",
    "    .agg(\n",
    "        mean_counts=pd.NamedAgg(column=counts_var, aggfunc='median'),\n",
    "        n=pd.NamedAgg(column='motif', aggfunc='count'),\n",
    "    )\n",
    "    .query('n >= 10')\n",
    "    .pivot_table(index=['mut_type_arrow', 'motif'], columns='ss_prediction', values='mean_counts')\n",
    "    .reset_index()\n",
    "    .dropna()\n",
    ")\n",
    "\n",
    "# Plot data\n",
    "nobs = data['mut_type_arrow'].value_counts()\n",
    "mut_types_to_plot = list(nobs[nobs > 1].index.values)\n",
    "data = data[data['mut_type_arrow'].isin(mut_types_to_plot)]\n",
    "g = sns.FacetGrid(data, col='mut_type_arrow',  col_wrap=4, sharex=False, sharey=False, height=2, aspect=1)\n",
    "g.map(sns.scatterplot, 'paired', 'unpaired', alpha=0.5, s=85)\n",
    "g.set_titles(\"{col_name}\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Analyze how much 3-mer motif medians differ between large segments of the genome (capturing trends related to global sequence context)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get data\n",
    "counts_var = 'truncated_actual_count'\n",
    "data = (\n",
    "    counts_df[counts_df['mut_class'] == 'synonymous']\n",
    "    .groupby(['mut_type_arrow', 'motif', 'nt_site_before_21555'], as_index=False)\n",
    "    .agg(\n",
    "        mean_counts=pd.NamedAgg(column=counts_var, aggfunc='median'),\n",
    "        n=pd.NamedAgg(column='motif', aggfunc='count'),\n",
    "    )\n",
    "    .query('n >= 10')\n",
    "    .sort_values('mean_counts', ascending=False)\n",
    "    .pivot_table(index=['mut_type_arrow', 'motif'], columns='nt_site_before_21555', values='mean_counts')\n",
    "    .reset_index()\n",
    "    .dropna()\n",
    ")\n",
    "\n",
    "# Plot data\n",
    "nobs = data['mut_type_arrow'].value_counts()\n",
    "mut_types_to_plot = list(nobs[nobs > 1].index.values)\n",
    "data = data[data['mut_type_arrow'].isin(mut_types_to_plot)]\n",
    "data.rename(columns={True : 'before', False : 'after'}, inplace=True)\n",
    "g = sns.FacetGrid(data, col='mut_type_arrow',  col_wrap=4, sharex=False, sharey=False, height=2, aspect=1.1)\n",
    "g.map(sns.scatterplot, 'before', 'after', alpha=0.5, s=85)\n",
    "g.set_titles(\"{col_name}\")\n",
    "#plt.suptitle('counts before or after nucleotide 21,555 in the genome', y=1.04)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Examine the correlation in 3mer motif patterns between CT and GA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get data\n",
    "ct_paired_data = counts_df[\n",
    "    (counts_df['mut_class'] == 'synonymous') &\n",
    "    (counts_df['ss_prediction'].isin(['paired'])) &\n",
    "    (counts_df['mut_type'] == 'CT')\n",
    "].copy()\n",
    "ct_paired_data['mut_type'] = 'CT_paired'\n",
    "\n",
    "ct_unpaired_data = counts_df[\n",
    "    (counts_df['mut_class'] == 'synonymous') &\n",
    "    (counts_df['ss_prediction'].isin(['unpaired'])) &\n",
    "    (counts_df['mut_type'] == 'CT')\n",
    "].copy()\n",
    "ct_unpaired_data['mut_type'] = 'CT_unpaired'\n",
    "\n",
    "ga_data = counts_df[\n",
    "    (counts_df['mut_class'] == 'synonymous') &\n",
    "    (counts_df['ss_prediction'].isin(['unpaired', 'paired'])) &\n",
    "    (counts_df['mut_type'] == 'GA')\n",
    "].copy()\n",
    "\n",
    "data = pd.concat([ct_paired_data, ct_unpaired_data, ga_data])\n",
    "\n",
    "# Compute the median counts of syn mutations for each motif\n",
    "medians_df = (\n",
    "    data[data['mut_class'] == 'synonymous']\n",
    "    .groupby(['mut_type', 'motif'], as_index=False)\n",
    "    .agg(\n",
    "        mean_counts=pd.NamedAgg(column=counts_var, aggfunc='median'),\n",
    "        n=pd.NamedAgg(column='motif', aggfunc='count'),\n",
    "    )\n",
    "    .query('n >= 10')\n",
    "    .sort_values('mean_counts', ascending=False)\n",
    "    .dropna()\n",
    ")\n",
    "medians_df['ref_motif'] = medians_df.apply(\n",
    "    lambda row: get_ref_motif(row['mut_type'], row['motif']), axis=1\n",
    ")\n",
    "\n",
    "# Plot data\n",
    "data = (\n",
    "    medians_df\n",
    "    .pivot_table(index='ref_motif', columns='mut_type', values='mean_counts')\n",
    "    #.dropna()\n",
    ")\n",
    "(fig, axs) = plt.subplots(ncols=2, sharey=True, figsize=[5.4,2.5])\n",
    "sns.scatterplot(x='CT_unpaired', y='GA', data=data, alpha=1.0, s=85, ax=axs[0])\n",
    "sns.scatterplot(x='CT_paired', y='GA', data=data, alpha=1.0, s=85, ax=axs[1])\n",
    "\n",
    "for (i, x) in enumerate(['CT_unpaired', 'CT_paired']):\n",
    "    r = data[x].corr(data['GA'])\n",
    "    axs[i].annotate(f'R = {round(r, 2)}', [0.05, 0.9], xycoords='axes fraction')\n",
    "\n",
    "axs[0].set(ylabel=r'G$\\rightarrow$A', xlabel=r'C$\\rightarrow$T (unpaired)', ylim=[0,940] , xlim=[0,4500])\n",
    "axs[1].set(xlabel=r'C$\\rightarrow$T (paired)', xlim=[0,1900])\n",
    "sns.despine()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Examine correlation between motif medians for counts from pre-Omicron clades vs. Omicron clades"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "g = sns.FacetGrid(\n",
    "    motif_medians.rename(columns={'median_counts_pre_omicron':'pre-Omicron', 'median_counts_omicron':'Omicron'}),\n",
    "    col='mut_type_arrow',  col_wrap=4, sharex=False, sharey=False, height=2, aspect=1\n",
    ")\n",
    "g.map(sns.scatterplot, 'pre-Omicron', 'Omicron', alpha=0.5, s=85)\n",
    "g.set_titles(\"{col_name}\")\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ncov-ab-escape",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
