{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Make a dataframe with curated mutation counts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Potential suggestions for updating Bloom repo:\n",
    "* add 3mer sequence context to `clade_founder_nts.csv`\n",
    "* update list of sites to ignore with sites from https://virological.org/t/issues-with-sars-cov-2-sequencing-data/473?\n",
    "* filter based on genome length and filter out non-human sequences?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in Python modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Identify sites that are conserved in all clade founders"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read in clade founder sequences, add a column giving a site's sequence context, then make a list of sites where the site, its codon, and its 3mer sequence context are conserved across all founders and identical to the Wuhan-Hu-1 reference sequence (ignoring the codon requirement for nonidentical sites)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of sites in genome: 29903\n",
      "Number of conserved sites: 28702\n"
     ]
    }
   ],
   "source": [
    "# Read in data\n",
    "fitness_results_dir = '../SARS2-mut-fitness/results_gisaid_2024-04-24'\n",
    "founder_df = pd.read_csv(os.path.join(fitness_results_dir, 'clade_founder_nts/clade_founder_nts.csv'))\n",
    "founder_df.sort_values(['clade', 'site'], inplace=True)\n",
    "\n",
    "# Get founder seqs\n",
    "founder_seq_dict = {}\n",
    "for (clade, data) in founder_df.groupby('clade'):\n",
    "    founder_seq_dict[clade] = ''.join(data['nt'])\n",
    "\n",
    "# For each row, get the site's 3mer motif in the corresponding founder sequence\n",
    "def get_motif(site, clade):\n",
    "    founder_seq = founder_seq_dict[clade]\n",
    "    return founder_seq[site-2:site+1]\n",
    "min_and_max_sites = [founder_df['site'].min(), founder_df['site'].max()]\n",
    "founder_df['motif'] = founder_df.apply(\n",
    "    lambda row: np.nan if row['site'] in min_and_max_sites \\\n",
    "        else get_motif(row['site'], row['clade']),\n",
    "    axis=1\n",
    ")\n",
    "\n",
    "# Add columns giving the reference codon and motif\n",
    "founder_df = founder_df.merge(\n",
    "    (\n",
    "        founder_df[founder_df['clade'] == '19A']\n",
    "        .rename(columns={'codon' : 'ref_codon', 'motif' : 'ref_motif'})\n",
    "    )[['site', 'ref_codon', 'ref_motif']], on='site', how='left'\n",
    ")\n",
    "\n",
    "# Identify sites where the codon and motif are conserved across all clade founders\n",
    "# by subsetting data to entries with identical codons/motifs to reference, then\n",
    "# identifying sites that still have entries for all clades\n",
    "data = founder_df[\n",
    "    (founder_df['codon'] == founder_df['ref_codon']) &\n",
    "    (founder_df['motif'] == founder_df['ref_motif'])\n",
    "]\n",
    "site_counts = data['site'].value_counts()\n",
    "nclades = len(founder_df['clade'].unique())\n",
    "conserved_sites = site_counts[site_counts == nclades].index\n",
    "founder_df['same_context_all_founders'] = founder_df['site'].isin(conserved_sites)\n",
    "founder_df['nt_site'] = founder_df['site']\n",
    "\n",
    "print('Number of sites in genome:', len(founder_df['site'].unique()))\n",
    "print('Number of conserved sites:', len(conserved_sites))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in and curate counts data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read in dataframe on actual and expected counts, and add columns with metadata."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in data\n",
    "counts_df = pd.read_csv(os.path.join(\n",
    "    fitness_results_dir,\n",
    "    'expected_vs_actual_mut_counts/expected_vs_actual_mut_counts.csv'\n",
    "))\n",
    "\n",
    "# Add metadata\n",
    "counts_df[['wt_nt', 'mut_nt']] = counts_df['nt_mutation'].str.extract(r'(\\w)\\d+(\\w)')\n",
    "counts_df['mut_type'] = counts_df['wt_nt'] + counts_df['mut_nt']\n",
    "\n",
    "def get_mut_class(row):\n",
    "    if row['synonymous']:\n",
    "        return 'synonymous'\n",
    "    elif row['noncoding']:\n",
    "        return 'noncoding'\n",
    "    elif '*' in row['mutant_aa']:\n",
    "        return 'nonsense'\n",
    "    elif row['mutant_aa'] != row['clade_founder_aa']:\n",
    "        return 'nonsynonymous'\n",
    "    else:\n",
    "        raise ValueError(row['mutant_aa'], row['clade_founder_aa'])\n",
    "\n",
    "counts_df['mut_class'] = counts_df.apply(lambda row: get_mut_class(row), axis=1)\n",
    "\n",
    "# Add column indicating if clade is pre-Omicron or Omicron\n",
    "pre_omicron_clades = [\n",
    "    '20A', '20B', '20C', '20E', '20G', '20H', '20I', '20J', '21C','21I', '21J'\n",
    "]\n",
    "counts_df['pre_omicron_or_omicron'] = counts_df['clade'].apply(\n",
    "    lambda x: 'pre_omicron' if x in pre_omicron_clades else 'omicron'\n",
    ")\n",
    "\n",
    "# Add column indicating if a site is before site 21,555\n",
    "counts_df['nt_site_before_21555'] = counts_df['nt_site'] < 21555\n",
    "\n",
    "# Add column indicating whether RNA sites from the Lan, 2022, Nature Comm. structure\n",
    "# are predicted to be paired, using code from Hensel, 2023, biorxiv\n",
    "filename = '../data/lan_2022/41467_2022_28603_MOESM11_ESM.txt'\n",
    "with open(filename) as f:\n",
    "    lines = [line.rstrip().split() for line in f]\n",
    "paired = np.array([[int(x[0]),int(x[4])] for x in lines[1:]])\n",
    "paired_dict = dict(zip(paired[:,0], paired[:,1]))\n",
    "def assign_ss_pred(site):\n",
    "    if site not in paired_dict:\n",
    "        return 'nd'\n",
    "    elif paired_dict[site] == 0:\n",
    "        return 'unpaired'\n",
    "    else:\n",
    "        return 'paired'\n",
    "counts_df['ss_prediction'] = counts_df['nt_site'].apply(lambda x: assign_ss_pred(x))\n",
    "\n",
    "# Add columns giving a site's motif relative to the clade founder\n",
    "# and the reference sequence\n",
    "counts_df = counts_df.merge(\n",
    "    founder_df[['nt_site', 'clade', 'motif', 'ref_motif']],\n",
    "    on = ['nt_site', 'clade'], how='left',\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create a dataframe with curated counts. We curate the data in the following ways:\n",
    "* only analyze sites that pass the above conservation criteria\n",
    "* ignore sites that are annotated as being masked in any clade of the UShER tree (`masked_in_usher == True`), are annotated for exclusion (`exclude == True`), or were identified to highly homoplastic by De Maio et al. (https://virological.org/t/issues-with-sars-cov-2-sequencing-data/473)\n",
    "\n",
    "Then, subset the dataframe to one row for each possible mutation, including the following columns:\n",
    "* `actual_count`: gives the mutation's count for `subset == all` from the above dataframe of counts\n",
    "* additional columns give actual counts for subsets of the data, such as geographical subsets (England vs. USA) or phylogenetic subsets (pre-Omicron vs. Omicron)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>nt_site</th>\n",
       "      <th>nt_mutation</th>\n",
       "      <th>clade_founder_nt</th>\n",
       "      <th>gene</th>\n",
       "      <th>clade_founder_codon</th>\n",
       "      <th>clade_founder_aa</th>\n",
       "      <th>mutant_codon</th>\n",
       "      <th>mutant_aa</th>\n",
       "      <th>aa_mutation</th>\n",
       "      <th>synonymous</th>\n",
       "      <th>...</th>\n",
       "      <th>ref_motif</th>\n",
       "      <th>expected_count</th>\n",
       "      <th>actual_count</th>\n",
       "      <th>count_terminal</th>\n",
       "      <th>count_non_terminal</th>\n",
       "      <th>mean_log_size</th>\n",
       "      <th>actual_count_England</th>\n",
       "      <th>actual_count_USA</th>\n",
       "      <th>actual_count_pre_omicron</th>\n",
       "      <th>actual_count_omicron</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>266</td>\n",
       "      <td>A266C</td>\n",
       "      <td>A</td>\n",
       "      <td>ORF1a;ORF1ab</td>\n",
       "      <td>ATG;ATG</td>\n",
       "      <td>M;M</td>\n",
       "      <td>CTG;CTG</td>\n",
       "      <td>L;L</td>\n",
       "      <td>M1L;M1L</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>GAT</td>\n",
       "      <td>22.594416</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>266</td>\n",
       "      <td>A266G</td>\n",
       "      <td>A</td>\n",
       "      <td>ORF1a;ORF1ab</td>\n",
       "      <td>ATG;ATG</td>\n",
       "      <td>M;M</td>\n",
       "      <td>GTG;GTG</td>\n",
       "      <td>V;V</td>\n",
       "      <td>M1V;M1V</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>GAT</td>\n",
       "      <td>177.066360</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>266</td>\n",
       "      <td>A266T</td>\n",
       "      <td>A</td>\n",
       "      <td>ORF1a;ORF1ab</td>\n",
       "      <td>ATG;ATG</td>\n",
       "      <td>M;M</td>\n",
       "      <td>TTG;TTG</td>\n",
       "      <td>L;L</td>\n",
       "      <td>M1L;M1L</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>GAT</td>\n",
       "      <td>51.769470</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.34657</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>267</td>\n",
       "      <td>T267A</td>\n",
       "      <td>T</td>\n",
       "      <td>ORF1a;ORF1ab</td>\n",
       "      <td>ATG;ATG</td>\n",
       "      <td>M;M</td>\n",
       "      <td>AAG;AAG</td>\n",
       "      <td>K;K</td>\n",
       "      <td>M1K;M1K</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>ATG</td>\n",
       "      <td>36.816650</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>267</td>\n",
       "      <td>T267C</td>\n",
       "      <td>T</td>\n",
       "      <td>ORF1a;ORF1ab</td>\n",
       "      <td>ATG;ATG</td>\n",
       "      <td>M;M</td>\n",
       "      <td>ACG;ACG</td>\n",
       "      <td>T;T</td>\n",
       "      <td>M1T;M1T</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>ATG</td>\n",
       "      <td>165.161750</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   nt_site nt_mutation clade_founder_nt          gene clade_founder_codon  \\\n",
       "0      266       A266C                A  ORF1a;ORF1ab             ATG;ATG   \n",
       "1      266       A266G                A  ORF1a;ORF1ab             ATG;ATG   \n",
       "2      266       A266T                A  ORF1a;ORF1ab             ATG;ATG   \n",
       "3      267       T267A                T  ORF1a;ORF1ab             ATG;ATG   \n",
       "4      267       T267C                T  ORF1a;ORF1ab             ATG;ATG   \n",
       "\n",
       "  clade_founder_aa mutant_codon mutant_aa aa_mutation  synonymous  ...  \\\n",
       "0              M;M      CTG;CTG       L;L     M1L;M1L       False  ...   \n",
       "1              M;M      GTG;GTG       V;V     M1V;M1V       False  ...   \n",
       "2              M;M      TTG;TTG       L;L     M1L;M1L       False  ...   \n",
       "3              M;M      AAG;AAG       K;K     M1K;M1K       False  ...   \n",
       "4              M;M      ACG;ACG       T;T     M1T;M1T       False  ...   \n",
       "\n",
       "   ref_motif expected_count actual_count  count_terminal count_non_terminal  \\\n",
       "0        GAT      22.594416            0               0                  0   \n",
       "1        GAT     177.066360            2               2                  0   \n",
       "2        GAT      51.769470            2               1                  1   \n",
       "3        ATG      36.816650            2               2                  0   \n",
       "4        ATG     165.161750            1               1                  0   \n",
       "\n",
       "  mean_log_size actual_count_England actual_count_USA  \\\n",
       "0       0.00000                    0                0   \n",
       "1       0.00000                    0                1   \n",
       "2       0.34657                    0                1   \n",
       "3       0.00000                    1                1   \n",
       "4       0.00000                    0                0   \n",
       "\n",
       "   actual_count_pre_omicron actual_count_omicron  \n",
       "0                         0                    0  \n",
       "1                         0                    2  \n",
       "2                         0                    2  \n",
       "3                         0                    2  \n",
       "4                         0                    1  \n",
       "\n",
       "[5 rows x 31 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Ignore sites that are masked or excluded in any clade of the UShER tree\n",
    "sites_to_ignore = list(counts_df[\n",
    "    (counts_df['masked_in_usher'] == True) |\n",
    "    (counts_df['exclude'] == True)\n",
    "]['nt_site'].unique())\n",
    "\n",
    "# Homoplastic sites from De Maio et al., which we will also ignore\n",
    "sites_to_ignore += [\n",
    "    187, 1059, 2094, 3037, 3130, 6990, 8022, 10323, 10741, 11074, 13408,\n",
    "    14786, 19684, 20148, 21137, 24034, 24378, 25563, 26144, 26461, 26681, 28077,\n",
    "    28826, 28854, 29700, 4050, 13402, 11083, 15324, 21575\n",
    "]\n",
    "\n",
    "# Aggregate counts across...\n",
    "# ... all clades for \"all\" subset\n",
    "ignore_cols = [\n",
    "    'expected_count', 'actual_count', 'count_terminal', 'count_non_terminal', 'mean_log_size',\n",
    "    'clade', 'pre_omicron_or_omicron'\n",
    "]\n",
    "groupby_cols = [\n",
    "    col for col in counts_df.columns.values\n",
    "    if col not in ignore_cols\n",
    "]\n",
    "curated_counts_df = counts_df[\n",
    "    (counts_df['nt_site'].isin(conserved_sites)) &\n",
    "    ~(counts_df['nt_site'].isin(sites_to_ignore)) &\n",
    "    (counts_df['subset'] == 'all')\n",
    "].groupby(groupby_cols, as_index=False).agg('sum', numeric_only=True)\n",
    "assert sum(curated_counts_df['nt_mutation'].duplicated(keep=False)) == 0\n",
    "\n",
    "# ... England or USA, and merge counts column with above dataframe\n",
    "subsets = ['England', 'USA']\n",
    "for subset in subsets:\n",
    "    subset_data = counts_df[\n",
    "        (counts_df['nt_site'].isin(conserved_sites)) &\n",
    "        ~(counts_df['nt_site'].isin(sites_to_ignore)) &\n",
    "        (counts_df['subset'] == subset)\n",
    "    ].groupby(groupby_cols, as_index=False).agg('sum', numeric_only=True)\n",
    "    assert sum(subset_data['nt_mutation'].duplicated(keep=False)) == 0\n",
    "    assert len(subset_data) == len(curated_counts_df)\n",
    "    curated_counts_df = curated_counts_df.merge(\n",
    "        (\n",
    "            subset_data\n",
    "            .rename(columns={'actual_count' : f'actual_count_{subset}'})\n",
    "        )[['nt_mutation', f'actual_count_{subset}']], on='nt_mutation'\n",
    "    )\n",
    "\n",
    "# ... pre-Omicron or Omicron clades, and merge counts column with above dataframe\n",
    "subsets = ['pre_omicron', 'omicron']\n",
    "for subset in subsets:\n",
    "    subset_data = counts_df[\n",
    "        (counts_df['nt_site'].isin(conserved_sites)) &\n",
    "        ~(counts_df['nt_site'].isin(sites_to_ignore)) &\n",
    "        (counts_df['subset'] == 'all') &\n",
    "        (counts_df['pre_omicron_or_omicron'] == subset)\n",
    "    ].groupby(groupby_cols, as_index=False).agg('sum', numeric_only=True)\n",
    "    assert sum(subset_data['nt_mutation'].duplicated(keep=False)) == 0\n",
    "    assert len(subset_data) == len(curated_counts_df)\n",
    "    curated_counts_df = curated_counts_df.merge(\n",
    "        (\n",
    "            subset_data\n",
    "            .rename(columns={'actual_count' : f'actual_count_{subset}'})\n",
    "        )[['nt_mutation', f'actual_count_{subset}']], on='nt_mutation'\n",
    "    )\n",
    "\n",
    "# Save curated counts to an output file\n",
    "assert sum(curated_counts_df['motif'] != curated_counts_df['ref_motif']) == 0\n",
    "assert len(curated_counts_df) == len(curated_counts_df['nt_mutation'].unique())\n",
    "curated_counts_df.drop(columns=['subset', 'exclude', 'masked_in_usher'], inplace=True)\n",
    "outfile = '../results/curated_mut_counts.csv'\n",
    "curated_counts_df.to_csv(outfile, index=False)\n",
    "\n",
    "curated_counts_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Summary statistics of mutations in datset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique muts:\n",
      "In the full dataset: 90621\n",
      "In the curated dataset: 84138\n"
     ]
    }
   ],
   "source": [
    "print('Number of unique muts:')\n",
    "print('In the full dataset:', len(counts_df['nt_mutation'].unique()))\n",
    "print('In the curated dataset:', len(curated_counts_df['nt_mutation'].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of curated mutations per category:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "mut_class\n",
       "nonsynonymous    61611\n",
       "synonymous       18042\n",
       "nonsense          3996\n",
       "noncoding          489\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('Number of curated mutations per category:')\n",
    "curated_counts_df['mut_class'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Use the dataframe with counts at all sites to make a list of gene boundaires"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>min_site</th>\n",
       "      <th>max_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ORF1a</td>\n",
       "      <td>266</td>\n",
       "      <td>13480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ORF1b</td>\n",
       "      <td>13481</td>\n",
       "      <td>21552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>S</td>\n",
       "      <td>21563</td>\n",
       "      <td>25381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ORF3a</td>\n",
       "      <td>25393</td>\n",
       "      <td>26217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>E</td>\n",
       "      <td>26245</td>\n",
       "      <td>26469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M</td>\n",
       "      <td>26523</td>\n",
       "      <td>27188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ORF6</td>\n",
       "      <td>27202</td>\n",
       "      <td>27384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ORF7a</td>\n",
       "      <td>27394</td>\n",
       "      <td>27755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ORF7b</td>\n",
       "      <td>27757</td>\n",
       "      <td>27884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ORF8</td>\n",
       "      <td>27894</td>\n",
       "      <td>28256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>N</td>\n",
       "      <td>28274</td>\n",
       "      <td>29530</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ORF10</td>\n",
       "      <td>29558</td>\n",
       "      <td>29671</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene  min_site  max_site\n",
       "4   ORF1a       266     13480\n",
       "5   ORF1b     13481     21552\n",
       "11      S     21563     25381\n",
       "6   ORF3a     25393     26217\n",
       "0       E     26245     26469\n",
       "1       M     26523     27188\n",
       "7    ORF6     27202     27384\n",
       "8   ORF7a     27394     27755\n",
       "9   ORF7b     27757     27884\n",
       "10   ORF8     27894     28256\n",
       "2       N     28274     29530\n",
       "3   ORF10     29558     29671"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get gene boundaires\n",
    "gene_boundaries_df = counts_df.groupby('gene', as_index=False).agg(\n",
    "    min_site = ('nt_site', 'min'),\n",
    "    max_site = ('nt_site', 'max'),\n",
    ")\n",
    "gene_boundaries_df['gene'].replace('ORF1a;ORF1ab', 'ORF1a', inplace=True)\n",
    "gene_boundaries_df['gene'].replace('ORF1ab', 'ORF1b', inplace=True)\n",
    "gene_boundaries_df = gene_boundaries_df[\n",
    "    ~(gene_boundaries_df['gene'].str.contains(';')) &\n",
    "    ~(gene_boundaries_df['gene'].isin(['noncoding']))\n",
    "].reset_index(drop=True).sort_values('min_site')\n",
    "\n",
    "# Save list to file\n",
    "outfile = '../results/gene_boundaries.csv'\n",
    "gene_boundaries_df.to_csv(outfile, index=False)\n",
    "\n",
    "gene_boundaries_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
